### PR TITLE
Frontend authorization docs for plugin authors

### PR DESCRIPTION
**Pull request by Joonpark13**
Signed-off-by: Joon Park <joonp@spotify.com>

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by Joonpark13** [comment]
Blocked by https://github.com/backstage/backstage/pull/10450

File: docs/permission/plugin-authors/05-frontend-authorization.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+Take, for example, the "Add" button in our todo list application. When a user clicks this button, the frontend makes a `POST` request to our backend's `/todos` route. If a user tries to add a todo but is not authorized, they will have no way of knowing this until they perform the action and are faced with an error. This is a poor user experience. We can do better by disabling the add button.
+
+**NOTE:** Placing frontend components behind authorization cannot take the place of placing your backend routes behind authorization. Authorization checks on the frontend should be used in _addition_ to the corresponding backend authorization, as an improvement to the user experience. If you do not place your backend route behind authorization, a malicious actor can still send a request to the route even if you disabled the corresponding frontend component.
```
Comment (joeporpeglia) [inline docs/permission/plugin-authors/05-frontend-authorization.md:11]: Seems like other docs use this format for notes/hints

```suggestion
> Note: Placing frontend components behind authorization cannot take the place of placing your backend routes behind authorization. Authorization checks on the frontend should be used in _addition_ to the corresponding backend authorization, as an improvement to the user experience. If you do not place your backend route behind authorization, a malicious actor can still send a request to the route even if you disabled the corresponding frontend component.
```

File: docs/permission/plugin-authors/05-frontend-authorization.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
++             {!loadingPermission && (
++               <Button disabled={!canAddTodo} variant="contained" onClick={handleAdd}>
++                 Add
++               </Button>
++             )}
+            </Box>
+...
+```
+
+Here we are using the [`usePermission` hook](https://backstage.io/docs/reference/plugin-permission-react.usepermission) to easily communicate with the permission policy and receive a decision on whether this user is authorized to create a todo list item.
```
Comment (joeporpeglia) [inline docs/permission/plugin-authors/05-frontend-authorization.md:69]:
```suggestion
Here we are using the [`usePermission` hook](https://backstage.io/docs/reference/plugin-permission-react.usepermission) to communicate with the permission policy and receive a decision on whether this user is authorized to create a todo list item.
```

File: docs/permission/plugin-authors/05-frontend-authorization.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+...
+```
+
+And now you should see that you are not able to create a todo item from the frontend!
+
+## Using `RequirePermission`
+
+More realistically, you probably don't even want to show the entire part of the UI that is used for todo list item creation if the user can't create a todo list item. For such cases, you can use the provided [`RequirePermission` component](https://backstage.io/docs/reference/plugin-permission-react.requirepermission):
```
Comment (joeporpeglia) [inline docs/permission/plugin-authors/05-frontend-authorization.md:92]:
```suggestion
Providing a disabled state can be a helpful signal to users, but there may be cases where hiding the element is preferred. For such cases, you can use the provided [`RequirePermission` component](https://backstage.io/docs/reference/plugin-permission-react.requirepermission):
```

**Review by joeporpeglia — APPROVED** [review]
Looks good! Just a few copy suggestions

File: docs/permission/plugin-authors/05-frontend-authorization.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    const title = useRef('');
+    const [key, refetchTodos] = useReducer(i => i + 1, 0);
+    const [editElement, setEdit] = useState<Todo | undefined>();
+
+-   const { loading: loadingPermission, allowed: canAddTodo } = usePermission({ permission: todoListCreate });
+...
+
+        <Grid container spacing={3} direction="column">
+-         <Grid item>
+-           <Typography variant="body1">Add todo</Typography>
```
Comment (freben) [inline docs/permission/plugin-authors/05-frontend-authorization.md:122]: could have saved yourselves a fair bit of diff churn if you'd had a separate `<TodoPageContents />` component that was all you rendered inside, while you were showing iterations of the surrounding stuff

**Review by freben — APPROVED** [review]

File: docs/permission/plugin-authors/05-frontend-authorization.md (hunk 4 above)
Comment (Rugvip) [reply docs/permission/plugin-authors/05-frontend-authorization.md:122]: Yep +1, this hurts the readability quite a bit. Always tricky to see what actually changed in a huge diff like this

File: docs/permission/plugin-authors/05-frontend-authorization.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+```diff
+...
+  import {
+    alertApiRef,
+    discoveryApiRef,
+    fetchApiRef,
+    useApi,
+  } from '@backstage/core-plugin-api';
++ import { usePermission } from '@backstage/plugin-permission-react';
++ import { todoListCreate } from '@internal/plugin-todo-list-common';
```
Comment (Rugvip) [inline docs/permission/plugin-authors/05-frontend-authorization.md:33]: Probably missed the `*Permission` suffix as part of the example package right? I feel it's very important for readability. What I mean is that this should be `todoListCreatePermission`, and that we should updated ALL the things! accordingly.

**Review by Rugvip — APPROVED** [review]
Nice 😁

I take it this is somewhat blocked by #9913? Perhaps this could be merged with a note that it's an upcoming component or something like that? #9913 won't make it into 1.2

File: docs/permission/plugin-authors/05-frontend-authorization.md (hunk 4 above)
Comment (joeporpeglia) [reply docs/permission/plugin-authors/05-frontend-authorization.md:122]: Good idea! Seems like a change worth making considering how many diffs we have throughout the tutorial

**Comment by Joonpark13** [comment]
We still plan on tackling these changes soon! Just haven't found the bandwidth yet.

**Review by Rugvip — APPROVED** [review]
Time to clean up the conflict and :shipit: right? The router changes are in

File: docs/permission/plugin-authors/05-frontend-authorization.md (hunk 4 above)
Comment (Joonpark13) [reply docs/permission/plugin-authors/05-frontend-authorization.md]: Done!

File: docs/permission/plugin-authors/05-frontend-authorization.md (hunk 5 above)
Comment (Joonpark13) [reply docs/permission/plugin-authors/05-frontend-authorization.md]: Done!

**Pull request merged**